### PR TITLE
39 next approver validation

### PR DESCRIPTION
**Pull request by tom08**

### CODE REVIEW

File: exdb/forms.py [hunk 1]
```diff
@@ -72,59 +74,62 @@ class Meta:
...
+             ValidationError(_('%(name)s events must have an audience') % {'name': name})),
+            (self.cleaned_data.get('type') and needs_verification and self.cleaned_data.get('attendance'),
+             ValidationError(_('%(name)s events cannot have an attendance') % {'name': name})),
+            (self.cleaned_data.get('type') and needs_verification and self.cleaned_data.get('start_datetime', min_dt) < self.when,
+             ValidationError(_('%(name)s events cannot happen in the past') % {'name': name})),
+            (needs_verification and self.cleaned_data.get('next_approver')
+                and not self.cleaned_data.get('next_approver').groups.filter(name='hs').exists(),
+             ValidationError(_('Supervisor must have permissions to approve and deny experiences'))),
+            (not needs_verification and not self.cleaned_data.get('conclusion'),
+             ValidationError(_('%(name)s events must have a conclusion') % {'name': name})),
```
Comment (enki211) [inline exdb/forms.py]: Submitting an empty experience gives this error, Without showing the user the conclusion field. Not really a big error, just an observation

**Comment by enki211** [comment]
Probably not for this pull request but for the UI overhaul. The edit and create pages display the list of errors with different styles

File: exdb/forms.py [hunk 2]
```diff
@@ -72,59 +74,62 @@ class Meta:
...
+                'attendance') or self.cleaned_data.get('attendance') < 1),
+             ValidationError(_('%(name)s events must have an attendance') % {'name': name})),
+            (self.cleaned_data.get('type') and not needs_verification and not self.cleaned_data.get('audience'),
+             ValidationError(_('%(name)s events must have an audience') % {'name': name})),
+            (self.cleaned_data.get('type') and needs_verification and self.cleaned_data.get('attendance'),
+             ValidationError(_('%(name)s events cannot have an attendance') % {'name': name})),
+            (self.cleaned_data.get('type') and needs_verification and self.cleaned_data.get('start_datetime', min_dt) < self.when,
+             ValidationError(_('%(name)s events cannot happen in the past') % {'name': name})),
+            (needs_verification and self.cleaned_data.get('next_approver')
+                and not self.cleaned_data.get('next_approver').groups.filter(name='hs').exists(),
```
Comment (shaneallgeier) [inline exdb/forms.py]: This should be updated to use the new `is_hallstaff()` method on user objects.

**Comment by tom08** [comment]
I'll probably leave the difference in formatting for the UI overhall.

**Comment by enki211** [comment]
Looks good

File: exdb/forms.py [hunk 3]
```diff
@@ -73,59 +75,62 @@ class Meta:
...
+            (not self.cleaned_data.get('type'), ValidationError(_('The type field is required'))),
+            (self.cleaned_data.get('type') and not self.approval_form and
+                not self.cleaned_data.get('next_approver') and needs_verification,
+             ValidationError(_('Please select the supervisor to review this experience'))),
+            (self.cleaned_data.get('start_datetime', max_dt) >= self.cleaned_data.get('end_datetime', min_dt),
+             ValidationError(_('Start time must be before end time'))),
+            ((self.cleaned_data.get('type') and not needs_verification) and (self.cleaned_data.get('start_datetime', max_dt) > self.when),
+             ValidationError(_('%(name)s experiences must have happened in the past') % {'name': name})),
+            ((self.cleaned_data.get('type') and not needs_verification) and (not self.cleaned_data.get(
+                'attendance') or self.cleaned_data.get('attendance') < 1),
```
Comment (shaneallgeier) [inline exdb/forms.py]:
```python
(not self.cleaned_data.get('attendance') or self.cleaned_data.get('attendance') < 1)
```
can be reduced to:
```python
self.cleaned_data.get('attendance', 0) < 1
```

File: exdb/forms.py [hunk 4]
```diff
@@ -73,59 +75,62 @@ class Meta:
...
-            raise ValidationError(_('%(name)s events cannot have an attendance') % {'name': ex_type.name})
-
-        # There are too many branches in this function; this is fixed on
-        # the approval validation branch.
-        if not ex_type.needs_verification and not self.cleaned_data.get('conclusion'):
-            raise ValidationError(_('%(name)s events must have a conclusion') % {'name': ex_type.name})
-
-        if ex_type.needs_verification and self.cleaned_data.get('conclusion'):
+        needs_verification = self.cleaned_data.get(
+            'type').needs_verification if self.cleaned_data.get('type') else False
```
Comment (shaneallgeier) [inline exdb/forms.py]: `(self.cleaned_data.get('type') and not needs_verification)` and `(self.cleaned_data.get('type') and needs_verification)` is repeated many times in the code below. Perhaps `needs_verification` should be a trinary variable (`True`/`False`/`None`) so the rest of the code is simpler.

For example:
```python
needs_verification = self.cleaned_data.get('type').needs_verification if self.cleaned_data.get('type') else False
...
(self.cleaned_data.get('type') and not needs_verification)
```
changes to:
```python
type = self.cleaned_data.get('type')
needs_verification = type.needs_verification if type else None
...
(needs_verification is False)
```

File: exdb/tests/integration_tests.py [hunk 5]
```diff
@@ -230,6 +230,12 @@ def test_experience_creation_spontaneous_no_conclusion(self):
         form = ExperienceSubmitForm(data, when=self.test_date)
         self.assertFalse(form.is_valid(), "Form should not be valid with no conclusion if it does not need approval")
 
+    def test_experience_creation_invalid_supervisor(self):
+        data = self.get_post_data((self.test_date + timedelta(days=1)), (self.test_date + timedelta(days=2)))
+        data['next_approver'] = self.clients['ra'].user_object.pk
+        form = ExperienceSubmitForm(data, when=self.test_date)
+        self.assertFalse(form.is_valid(), "Form should not be valid if next_approver user does not have 'hs' group")
```
Comment (shaneallgeier) [inline exdb/tests/integration_tests.py]: It's not really called `hs` anymore. You might just say "Form should not be valid if next_approver user isn't hallstaff".

**Comment by shaneallgeier** [comment]
I approve.

**Comment by enki211** [comment]
Looks good

**Pull request merged**